lib/checkother.cpp

```diff
@@ -485,6 +485,12 @@ void CheckOther::checkRedundantAssignment()
                 else
                     start = tok->findExpressionStartEndTokens().second->next();
 
+                // Is this a variable declaration with initialization?
+                const Token* defineAndInitTok = start->tokAt(-3);
+                bool isInitialization = false;
+                if (Token::Match(defineAndInitTok, "; %var% =") && defineAndInitTok->linenr() == start->linenr())
```

Comment:
I'd suggest that you check if the assignment is preceded by the variable declaration.. something like:

    const Token *defineAndInitTok = start->tokAt(-4);
    if (Token::Match(defineAndInitTok, "%var% ; %var% =") &&
        defineAndInitTok->variable() &&
        defineAndInitTok->variable() == defineAndInitTok->tokAt(2)->variable() &&
        defineAndInitTok->variable()->nameToken() == defineAndInitTok)

I guess that you can also check if this is true: `defineAndInitTok->column() == defineAndInitTok->tokAt(2)->column()`

Reply:
hmm.. or maybe a utility function in astutils would be good to have. I can look into that if you like. I have the feeling we have such logic here and there.

Reply:
hmm if I change the check as you suggested all redundantVarAssignment tests fail ;)
One question still remains: is the new check of type "style" or "warning"
It's a bit of both actually...

Reply:
> 
> 
> I'd suggest that you check if the assignment is preceded by the variable declaration.. something like:
> 
> ```
> const Token *defineAndInitTok = start->tokAt(-4);
> if (Token::Match(defineAndInitTok, "%var% ; %var% =") &&
>     defineAndInitTok->variable() &&
>     defineAndInitTok->variable() == defineAndInitTok->tokAt(2)->variable() &&
>     defineAndInitTok->variable()->nameToken() == defineAndInitTok)
> ```
> 
> I guess that you can also check if this is true: `defineAndInitTok->column() == defineAndInitTok->tokAt(2)->column()`

The check shouldn't only detect these kind of intialization and assignment situations. The check should also warn the user if you have something like this:

int x=0;
/*do something without using x*/
for(x=0;x<....)

Reply:
Imho a initialization looks like this:

    int x = 0;

With your code I fear that you will say that this is initialization also:

    somecode; var = 0;

Reply:
> hmm if I change the check as you suggested all redundantVarAssignment tests fail ;)

I might have made some mistake. But well I do think it would be good to determine if there is initialization better.

Reply:
In Cppcheck master we have this code:

                // Do not warn about redundant initialization when rhs is trivial
                // TODO : do not simplify the variable declarations
                if (Token::Match(tok->tokAt(-3), "%var% ; %var% =") && tok->previous()->variable() && tok->previous()->variable()->nameToken() == tok->tokAt(-3) && tok->tokAt(-3)->linenr() == tok->previous()->linenr()) {
                    bool trivial = true;
                    visitAstNodes(tok->astOperand2(),
                    [&](const Token *rhs) {
                        if (Token::Match(rhs, "%str%|%num%|%name%"))
                            return ChildrenToVisit::op1_and_op2;
                        if (rhs->str() == "(" && !rhs->previous()->isName())
                            return ChildrenToVisit::op1_and_op2;
                        trivial = false;
                        return ChildrenToVisit::done;
                    });
                    if (trivial)
                        continue;
                }

Could you change that?